index.js

```diff
@@ -441,11 +447,21 @@ function browserslist(queries, opts) {
   return result
 }
 
+function parseQueries(queries) {
+  var cacheKey = JSON.stringify(queries)
```

Comment:
I think a better way is to use `configPath` as cache key and not config content.

With `configPath` as cache key we don’t need to parse config, etc.

Reply:
There is not always a `configPath`, because we can get config from `opts.queries`, `env.BROWSERSLIST`, or the fallback to `browserslist.defaults|

Besides, there's no repeated parsing, because `parsePackageOrReadConfig` now had its own cache

Reply:
Hm. Yes, you are right.

In this case, the path is great. Let’s avoid `ctx.path` changes and we are ready to merge.

Reply:
Sorry, still don't get what you mean by "avoid `ctx.path` changes". Something like this?

```diff
-var cacheKey = JSON.stringify([queries, context])
+var cacheKey = JSON.stringify([queries, {
+  ignoreUnknownVersions: context.ignoreUnknownVersions,
+  dangerousExtend: context.dangerousExtend,
+  mobileToDesktop: context.mobileToDesktop,
+  env: context.env,
+  path: needsPath ? context.path : undefined
+}])
```

Reply:
https://github.com/browserslist/browserslist/pull/862#discussion_r1888380971

Reply:
For instance, add `ctx.cachePath` or something like this and check every time when we use `ctx.path`

Reply:
I only see two ways to, at the same time:

1. Keep `path` in context
2. Remove `path` from cache key
3. Avoid cloning `context` in multiple places, which is verbose and error-prone in es5

One is the `definePropery` trick I just pushed, the other is a sprinkle of TS to enforce `usesPath` on queries that need `{ path: string }`. I'm probably missing the point, but see if you find any of the options suitable.

Reply:
Hey @ai, could you take a look?

Reply:
Now I get it why you deleted `ctx.path`. I initially thought that it is to disable cache, but it is because some query was not subscribed to the path?

Reply:
OK, since I still don't understand the confusion, here's the most detailed and boring exlainer to avoid any miscommunication:

1. `context` object is passed to all `QUERIES[*].select` _and_ used as the cache key
2. `context` contains `opts.path`, which is the path of the file currently processed
3. Putting `path` in cache key causes resolve cache to become useless for processing multiple files
4. Only two queries, `browserslist config` and `extends` actually use `path`

Therefore, we remove `path` from `context` unsell config contains any of the two path-dependent queries. 

We accept two risks:

1. Reading `context.path` from a query not marked with `needsPath` will fail, and probably cause debug trouble
2. Resolving configs with `browserslist config` or `extends` queries will be de-optimized and mostly bypass cache.

I could create an issue for risk 2 if you want, fixing (non-`dangerous`) `extends` shouldn't be too hard.

Reply:
After the holiday I will do I final review for this PR.

You can add “non-dangerous extends” support to this PR if you have a free time before it.

Dangerous extends is much rare case, so it is OK to keep it de-optimized. Non-dangerous extends is rare too, but I know that some people use it.